clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -522,78 +522,72 @@ static Expected<StringRef> convertSPIRVToIR(StringRef Filename,
   return *TempFileOrErr;
 }
 
-// Update sycl-post-link options based on target triple.
-static void updateCmdArgs(SmallVector<StringRef, 8> &CmdArgs,
-                          llvm::Triple Triple) {
-  // Get an argument in CmdArgs that contains Str. If there is no such
-  // argument, an empty argument is returned
-  auto getArg = [&](const StringRef &Str) {
-    for (auto Arg : CmdArgs)
-      if (Arg.contains(Str))
-        return Arg;
-    return StringRef("");
-  };
-  // Add a new argument Arg to CmdArgs if not present already.
-  auto addArg = [&](const StringRef &Arg) {
-    if (getArg(Arg).empty())
-      CmdArgs.push_back(Arg);
-  };
-  // Replace an argument in CmdArgs that contains Str with NewArg. If no such
-  // argument is present, add the NewArg to CmdArgs.
-  auto replaceOrAddArg = [&](const StringRef &NewArg, const StringRef &Str) {
-    for (auto &Arg : CmdArgs)
-      if (Arg.contains(Str)) {
-        Arg = NewArg;
-        return;
-      }
-    CmdArgs.push_back(NewArg);
-  };
-  // Remove argument containing Str from CmdArgs.
-  auto removeArg = [&](const StringRef &Str) {
-    CmdArgs.erase(
-        std::remove_if(CmdArgs.begin(), CmdArgs.end(),
-                       [&](StringRef Arg) { return Arg.contains(Str); }),
-        CmdArgs.end());
-  };
-
-  // specialization constants processing.
-  bool IsAOTGPU = Triple.isNVPTX() || Triple.isAMDGCN() || Triple.isSPIRAOT();
-  if (!IsAOTGPU)
-    replaceOrAddArg("-spec-const=native", "-spec-const");
-  else
-    replaceOrAddArg("-spec-const=emulation", "-spec-const");
-
-  // -emit-only-kernels-as-entry-points is set by the user and is enabled only
-  // for Intel targets.
-  auto EmitOnlyKernelsAsEntryPointsArg =
-      getArg("-emit-only-kernels-as-entry-points");
-  if ((!EmitOnlyKernelsAsEntryPointsArg.empty()) && !Triple.isNVPTX() &&
-      !Triple.isAMDGPU())
-    addArg("-emit-only-kernels-as-entry-points");
-  else
-    removeArg("-emit-only-kernels-as-entry-points");
-
-  if (!(Triple.isAMDGCN()))
-    addArg("-emit-param-info");
-
-  if (Triple.isNVPTX() || Triple.isAMDGCN())
-    addArg("-emit-program-metadata");
-
-  if (Triple.isSPIROrSPIRV()) {
-    addArg("-symbols");
-    addArg("-emit-exported-symbols");
-    addArg("-split-esimd");
-    addArg("-lower-esimd");
-  }
-
-  // Here, IsAOT includes x86_64 device as well.
-  bool IsAOT =
-      IsAOTGPU || Triple.getSubArch() == llvm::Triple::SPIRSubArch_x86_64;
-  auto GenDeviceImageArg = getArg("-generate-device-image-default-spec-consts");
-  if ((!GenDeviceImageArg.empty()) && IsAOT)
-    addArg("-generate-device-image-default-spec-consts");
+// Add any sycl-post-link options that rely on a specific Triple in addition
+// to user supplied options.
+// NOTE: Any changes made here should be reflected in the similarly named
+// function in clang/lib/Driver/ToolChains/Clang.cpp.
+static void
+getTripleBasedSYCLPostLinkOpts(const ArgList &Args,
```

Comment:
Sorry I might have missed something, but can you remind me why we need two functions, one that cares about the triple and one that doesn't? It seem like it would be simpler if we had a single function because then we won't need to rely on things the previous one did.

Reply:
It's due to how clang-linker-wrapper processes images. clang-linker-wrapper receives a set of images each with its own target triple. So, we do not have a singular target triple at the call site of clang-linker-wrapper. The options which do not depend on the target triple can be computed once and passed into the linker wrapper. The options that depend on the target triple will need to be computed for each device image group ( group corresponds to a target triple) inside the clang-linker-wrapper. There are few other ways to implement this. one option will be to pass all required information to the clang-linker-wrapper and compute all of them in one place. Passing in all the required information seemed overkill. Current approach is agreeable, I Hope.

Thanks

Reply:
assuming it doesnt get too much more complex i think it's okay